### PR TITLE
Instrument combine operators query execution code with thread cpu time

### PR DESCRIPTION
**Pull request by mqliang**
## Description
Instrument the query execution code in combine operators to measure the thread cpu time. The measurement is disabled by default and can be enabled using an instance level config.

Instrumentation is conducted at BaseCombineOperator, so all combine operator extends BaseCombineOperator will get the instrumentation(unless it override the getNextBlock() function) 

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release.

If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text

## Documentation
If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document

### CODE REVIEW

**Comment by mqliang** [comment]
@siddharthteotia

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ */
+package org.apache.pinot.core.query.request.context;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class ThreadTimer {
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:27]: Please add javadoc

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java [hunk 2]
```diff
@@ -89,12 +92,28 @@ protected IntermediateResultsBlock getNextBlock() {
...
+
           processSegments(threadIndex);
+
+          processThreadTimer.stop();
+          totalWorkerTime.addAndGet(processThreadTimer.getThreadTime());
         }
       });
     }
 
+    ThreadTimer mergeThreadTimer = new ThreadTimer();
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java:106]: I don't think we should instrument the merge part this way.

We should do that in InstanceResponseOperator and InstanceResponseBlock because that is the single main thread and all the merge work is happening under that thread. 

So we start the main thread timer in InstanceResponseOperator and stop the main thread timer in InstanceResponseBlock's constructor and then add up the timer received from merged intermediate result block. This will cover both single threaded and multi-threaded code.

**Comment by siddharthteotia** [comment]
We should emit the metric like other server metrics and also see if the measurement can be made configurable via instance level config.

**Comment by mqliang** [comment]
> We should emit the metric like other server metrics and also see if the measurement can be made configurable via instance level config.

This is a query level metric. If we emit it as BrokerMetrics/BrokerGauge, the number of metrics will explode. But we can expose it as a part of query response metadata, just like numServersQueried/numDocsScanned).

**Comment by siddharthteotia** [comment]
> > We should emit the metric like other server metrics and also see if the measurement can be made configurable via instance level config.
> 
> This is a query level metric. If we emit it as BrokerMetrics/BrokerGauge, the number of metrics will explode. But we can expose it as a part of query response metadata, just like numServersQueried/numDocsScanned).

I don't follow. I am not sure how BrokerMetric or BrokerGauge is involved here in the server context.  From the server we should emit this query level metric and we should also include this as part of data table metadata (numDocsScanned etc) that is sent to the broker.

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  private static final ThreadMXBean MX_BEAN = ManagementFactory.getThreadMXBean();
+  private static final boolean IS_THREAD_CPU_TIME_SUPPORTED = MX_BEAN.isThreadCpuTimeSupported();
+  private static final long NANOS_IN_MILLIS = 1;
```
Comment (fx19880617) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:30]: this should be 10^6?

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    if (_startTimeMs == -1 || _endTimeMs == -1) {
+      return 0;
+    }
+    return _endTimeMs - _startTimeMs;
+  }
+
+  static {
+    LOGGER.info("Thread cpu time measurement supported: {}", IS_THREAD_CPU_TIME_SUPPORTED);
+  }
+}
```
Comment (fx19880617) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:61]: nit: newline

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java [hunk 5]
```diff
@@ -44,6 +44,7 @@
   String REQUEST_ID_METADATA_KEY = "requestId";
   String NUM_RESIZES_METADATA_KEY = "numResizes";
   String RESIZE_TIME_MS_METADATA_KEY = "resizeTimeMs";
+  String THREAD_TIME_MS = "threadTimeMs";
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java:47]: Change it to micro seconds please

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java [hunk 6]
```diff
@@ -69,6 +69,7 @@
   private boolean _numGroupsLimitReached;
   private int _numResizes;
   private long _resizeTimeMs;
+  private long _threadTime;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java:72]: suggest: `threadCpuTimeMicros` instead of `threadTime` (Hint: Always add the units in variable name)

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java [hunk 7]
```diff
@@ -53,15 +55,16 @@
...
-  protected int _numThreads;
   // Use a Phaser to ensure all the Futures are done (not scheduled, finished or interrupted) before the main thread
   // returns. We need to ensure this because the main thread holds the reference to the segments. If a segment is
   // deleted/refreshed, the segment will be released after the main thread returns, which would lead to undefined
   // behavior (even JVM crash) when processing queries against it.
   protected final Phaser _phaser = new Phaser(1);
-  protected Future[] _futures;
   // Use a _blockingQueue to store the per-segment result
   private final BlockingQueue<IntermediateResultsBlock> _blockingQueue;
+  private final AtomicLong totalWorkerTime = new AtomicLong(0);
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java:65]: Suggest: `totalWorkerThreadCpuTimeMicros` or `totalThreadCpuMicros` or `execCpuTimeMicros` . Definitely include units

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java [hunk 8]
```diff
@@ -89,12 +92,28 @@ protected IntermediateResultsBlock getNextBlock() {
       _futures[i] = _executorService.submit(new TraceRunnable() {
         @Override
         public void runJob() {
+          ThreadTimer processThreadTimer = new ThreadTimer();
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java:95]: Suggest `executionThreadTimer`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java [hunk 9]
```diff
@@ -89,12 +92,28 @@ protected IntermediateResultsBlock getNextBlock() {
...
+          processThreadTimer.start();
+
           processSegments(threadIndex);
+
+          processThreadTimer.stop();
+          totalWorkerTime.addAndGet(processThreadTimer.getThreadTime());
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java:101]: Maybe we can introduce another API `ThreadTimer.stopAndGet()`

File: pinot-core/src/main/java/org/apache/pinot/core/operator/streaming/StreamingSelectionOnlyCombineOperator.java [hunk 10]
```diff
@@ -47,6 +47,7 @@
 
 /**
  * Combine operator for selection only streaming queries.
+ * TODO: extends StreamingSelectionOnlyCombineOperator from BaseCombineOperator.
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/streaming/StreamingSelectionOnlyCombineOperator.java:50]:
```suggestion
 * TODO: extend StreamingSelectionOnlyCombineOperator from BaseCombineOperator.
```

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 1 above)
Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:27]: +1

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import org.slf4j.LoggerFactory;
+
+
+public class ThreadTimer {
+  private static final ThreadMXBean MX_BEAN = ManagementFactory.getThreadMXBean();
+  private static final boolean IS_THREAD_CPU_TIME_SUPPORTED = MX_BEAN.isThreadCpuTimeSupported();
+  private static final long NANOS_IN_MILLIS = 1;
+  private static final Logger LOGGER = LoggerFactory.getLogger(ThreadTimer.class);
+
+  private long _startTimeMs = -1;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:33]: We need microseconds, not milliseconds

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 3 above)
Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:30]: Make it MICROS_IN_MILLIS

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 1 above)
Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:27]: Make this class configurable via instance configuration. If the instance configuration asks not to measure this, then do as appropriate in in the class below, with as minimal an overhead as possible. Even though we have measured the overhead to be near-zero in high throughput use cases, it is always useful to have a mechanism to turn it off in production when we first introduce it. It will also help us roll out the feature slowly in specific clusters or data centers

Eventually, we will get rid of the config and make it standard.

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 11 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:33]: Also introduce a static method that detects whether the underlying JVM supports getCurrentThreadCpuTme. If that is not supported (or if thread cpu time is not configured), then the overhead should be close to nothing.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java (hunk 5 above)
Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java:47]: Why should this be changed to microseconds?
The jmx api documentation says it returns nanosecond precision (but may not be necessarily accurate). But we can't treat it as microseconds.

For example, if the returned value is 
1000000. It is 1000 microseconds. If we treat the returned value as microseconds, then we are increasing the value by 1000 times making it 1000ms.

**Comment by mcvsubbu** [comment]
Just to clarify:
The value returned is in nanos. Initially in the prototype I had this converted to milli seconds, and most values turned out to be 0 when dividing nanos by 1_000_000 to get millis. So I changed it to log the values in nanos, and found that the resolution was much better (i;.e. we get non-zero numbers). It is ok to convert to microseconds. Or just leave it at nanoseconds. But converting to millis will not work

**Comment by mcvsubbu** [comment]
Let us settle to use nanosecond resolution.

**Comment by fx19880617** [comment]
> Just to clarify:
> The value returned is in nanos. Initially in the prototype I had this converted to milli seconds, and most values turned out to be 0 when dividing nanos by 1_000_000 to get millis. So I changed it to log the values in nanos, and found that the resolution was much better (i;.e. we get non-zero numbers). It is ok to convert to microseconds. Or just leave it at nanoseconds. But converting to millis will not work

Agreed. I also prefer nano resolution here.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java (hunk 2 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java]: If we instrument in that way (start the main thread timer in InstanceResponseOperator), then the total time will account an extra processing thread time interval:

example:
t0: InstanceResponseOperator call InstanceResponseBlock's constructor, which will call _operator.nextBlock();
t2: process segments in parallel. let's say we have 5 thread processing segments in parallel. Let's assume all the 5 thread start and end at the same time.
t3: processing end, start merge
t4: merge end.

NOTE: since time spending on constructor is negligible, let's assume t0=t2.

Then, expected_totalTime = processing_time + merge_time =  (t3-t2)*5 + (t4-t3). However, if start the main thread timer in InstanceResponseOperator, actual_totalTime = t4-t0+(t3-t2)*5 = t4-t2+(t3-t2)*5

actual_totalTime - expected_totalTime = (t4-t2)-(t4-t3) = t3-t2, in other word, there is an extra processing thread time get accounted.

I will let @mcvsubbu chime in to make a double check.

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 1 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:30]: done

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 3 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 4 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java (hunk 5 above)
Comment (mqliang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java (hunk 7 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java (hunk 6 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/IntermediateResultsBlock.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java (hunk 8 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java (hunk 9 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 11 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java [hunk 12]
```diff
@@ -317,6 +317,9 @@
     public static final String ACCESS_CONTROL_FACTORY_CLASS = "pinot.server.admin.access.control.factory.class";
     public static final String DEFAULT_ACCESS_CONTROL_FACTORY_CLASS =
         "org.apache.pinot.server.api.access.AllowAllAccessFactory";
+
+    public static final String CONFIG_OF_ENABLE_THREAD_CPU_TIME_INSTRUMENT = "pinot.server.instance.enableThreadCpuTimeInstrument";
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:321]: (nit) rename this to pinot.server.instance.enableThreadCpuTimeMeasurement

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java [hunk 13]
```diff
@@ -65,6 +65,7 @@
   SEGMENT_DOWNLOAD_FAILURES("segments", false),
   NUM_RESIZES("numResizes", false),
   RESIZE_TIME_MS("resizeTimeMs", false),
+  THREAD_CPU_TIME_NS("threadCpuTimeNs", false),
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java:68]: should be in `ServerGauge` not `ServerMeter`. Also looks like `RESIZE_TIME` is in the wrong place. Should be a gauge. You can fix that in a separate PR

File: pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java [hunk 14]
```diff
@@ -317,6 +317,9 @@
...
+    public static final boolean DEFAULT_ENABLE_THREAD_CPU_TIME_INSTRUMENT = true;
```
Comment (mcvsubbu) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java:322]: default should be false, at least for now

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java [hunk 15]
```diff
@@ -41,6 +41,9 @@
   public InstanceResponseBlock(IntermediateResultsBlock intermediateResultsBlock) {
     try {
       _instanceResponseDataTable = intermediateResultsBlock.getDataTable();
+      long totalThreadCpuTimeNs = intermediateResultsBlock.getThreadCpuTimeNs();
+      _instanceResponseDataTable.getMetadata().put(DataTable.THREAD_CPU_TIME_NS_METADATA_KEY,
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java:45]: Do we want to hold off on this until we also add the serialization part?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java]: As discussed offline, we should move all the single threaded instrumentation to InstanceResponseOperator

```
 @Override
  protected InstanceResponseBlock getNextBlock() {
    ThreadTimer mainThreadTimer = new ThreadTimer();
   
    mainThreadTimer.start();

    IntermediateResultsBlock intermediateResultsBlock = (IntermediateResultsBlock) _operator.nextBlock();
    InstanceResponseBlock instanceResponseBlock = new InstanceResponseBlock(intermediateResultsBlock);
    DataTable dataTable = instanceResponseBlock.getInstanceResponseDataTable();
   
    mainThreadTimer.stop();
   
    long totalThreadCpuTimeNs = intermediateResultsBlock.getThreadCpuTimeNs() + mainThreadTimer.getThreadTimeNs();
   
    dataTable.getMetadata().put(DataTable.THREAD_CPU_TIME_NS_METADATA_KEY, String.valueOf(totalThreadCpuTimeNs));

    return instanceResponseBlock
  }
```

This way all single threaded query execution (except for planning) code's cpu time is covered without having to do it in multiple places.

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/combine/BaseCombineOperator.java]: cc @mcvsubbu @mqliang

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java (hunk 15 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java:45]: I agree. Let's send the value from server to broker in the DataTable metadata (or whatever) only after we do the follow-up of measuring the data table serialization cost. cc @mqliang

Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java:45]: If we don't set the value as DataTable metadata here, then it's impossible to get the value in QueryScheduler.java, in other words, we should delete the ServerGauge emitting code also. I think at this moment, we can just put a log line here, and deleting the change in QueryScheduler.java. 

Another workaround is: add the value as DataTable metadata here, but in QueryScheduler.java, after writing log and emitting ServerGauge, `dataTable.getMetadata().remove(DataTable.THREAD_CPU_TIME_NS_METADATA_KEY)` so that this value will NOT be send back to broker (and add a TODO stating remove it when we have data table serialization cost measuring).

Any preference between the two option?

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java (hunk 13 above)
Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java:68]: Thinking more about this

We have ServerMeter, ServerGauge, ServerTimer and ServerQueryPhase. 

- This metric doesn't fit into ServerQueryPhase since we are not introducing a new phase.
- Similarly, ServerMeter is also not right as it is not used for emitting duration related metrics.

So between ServerGauge and ServerTimer, I am thinking we should use ServerTimer. Currently it is used for freshness lag and netty server response latency. We can introduce a new timer for CPU time measurements and then use `serverMetrics.addTimedTableValue(tableNameWithType, ServerTimer.THREAD_CPU_TIME ....)`

Now for the naming, we currently use `THREAD_CPU_TIME_NS`. Right now our focus is on two main costs for query execution

- Execution for segment level and combine - this PR
- DataTable serialization - next PR

For sending to the broker, it can add up both, but I feel we should have different metric names for cpu timers for the server to emit both separately. So, something like EXECUTION_THREAD_CPU_TIME_NS and RESPONSE_SERIALIZATION_CPU_TIME_NS. 

Secondly, in future we might decide to emit the metrics for let's say pruning, physical planning etc. Server can continue to add up everything and send to broker, but I think it should still emit individually for which we need separate names

@mcvsubbu @mqliang  what do you think ?

Comment (mcvsubbu) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java:68]: Agreed with changing the name to execution thread time.
I think the phase time just comes out as a gauge, so it is the same thing. Since this is not a separate query phase we are monitoring, I would leave it as a gauge.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java (hunk 15 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java:45]: I think logging should still be done in QueryScheduler since that's where all the metrics of a query are logged together in one line once the processing is done. So, we shouldn't log it in a new place now imo. 

As you suggested, I think we can do a put() here, retrieve the metric in QueryScheduler from dataTable metadata and then make sure to do a remove() before  `byte[] responseData = serializeDataTable(queryRequest, dataTable);` is called. Use the retrieved value for adding to table gauge and writing the log. For the remove(), add a TODO to revisit this when follow-up work of data table serialization cost measurement is done. Should be fine since we are going to do the follow-up immediately after this.

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java (hunk 13 above)
Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java:68]: @mqliang ^^, let's change the metric name.

**Comment by siddharthteotia** [comment]
@mqliang , can you please rebase this PR since https://github.com/apache/incubator-pinot/pull/6672 has been merged

**Comment by mqliang** [comment]
rebased and all comments addressed.

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+  private static final boolean IS_CURRENT_THREAD_CPU_TIME_SUPPORTED = MX_BEAN.isCurrentThreadCpuTimeSupported();
+  private static final Logger LOGGER = LoggerFactory.getLogger(ThreadTimer.class);
+  private static boolean IS_THREAD_CPU_TIME_MEASUREMENT_ENABLED;
+  private long _startTimeNs = -1;
+  private long _endTimeNs = -1;
+
+  public ThreadTimer() {
+  }
+
+  public static boolean isCurrentThreadCpuTimeSupported() {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:41]: remove this method, we should not need it.

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    stop();
+    return getThreadTimeNs();
+  }
+
+  static {
+    LOGGER.info("Current thread cpu time measurement supported: {}", IS_CURRENT_THREAD_CPU_TIME_SUPPORTED);
+  }
+
+  public static void setThreadCpuTimeMeasurementEnabled(boolean enable) {
+    IS_THREAD_CPU_TIME_MEASUREMENT_ENABLED = enable;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:74]: In this method, call ManagementFactory.getMxBean, and then decide once and for all whether we are going to measure thread cpu time or not, and store it in a single boolean variable, to be checked every time start/stop is invoked.

**Review by mcvsubbu — COMMENTED** [review]
Enable thread time measurement in integration tests

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java (hunk 13 above)
Comment (mqliang) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java]: Should we also make NUM_RESIZES as ServerGauge? Or just RESIZE_TIME_MS?

Comment (mcvsubbu) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java]: Only RESIZE_TIME

File: pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java [hunk 18]
```diff
@@ -65,7 +65,8 @@ public void setUp() {
   public void testSelectionOnlyCombineOperator() {
     List<Operator> operators = getOperators();
     SelectionOnlyCombineOperator combineOperator = new SelectionOnlyCombineOperator(operators,
-        QueryContextConverterUtils.getQueryContextFromPQL("SELECT * FROM table"), _executorService, TIMEOUT_MS);
+        QueryContextConverterUtils.getQueryContextFromPQL("SELECT * FROM table"),
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java:68]: This change doesn't seem to be related to the PR

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerGauge.java [hunk 19]
```diff
@@ -41,7 +41,7 @@
   REALTIME_OFFHEAP_MEMORY_USED("bytes", false),
   REALTIME_SEGMENT_NUM_PARTITIONS("realtimeSegmentNumPartitions", false),
   LLC_SIMULTANEOUS_SEGMENT_BUILDS("llcSimultaneousSegmentBuilds", true),
-
+  EXECUTION_THREAD_CPU_TIME_NS("executionThreadCpuTimeNs", false),
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerGauge.java:44]: We should pass the unit as the first argument.

**Review by siddharthteotia — APPROVED** [review]

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerGauge.java (hunk 19 above)
Comment (mqliang) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerGauge.java]: done

File: pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java (hunk 18 above)
Comment (mqliang) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java:68]: done

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java (hunk 17 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java (hunk 15 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseBlock.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java (hunk 12 above)
Comment (mqliang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/CommonConstants.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java (hunk 13 above)
Comment (mqliang) [reply pinot-common/src/main/java/org/apache/pinot/common/metrics/ServerMeter.java]: done

**Comment by mqliang** [comment]
conflict with master since https://github.com/apache/incubator-pinot/pull/6690 get merged. Rebased and updated.

File: pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java (hunk 18 above)
Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java:68]: I meant this change can be undone right?

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java [hunk 20]
```diff
@@ -77,6 +77,7 @@
   private PlanMaker _planMaker;
   private long _defaultTimeOutMs = CommonConstants.Server.DEFAULT_QUERY_EXECUTOR_TIMEOUT_MS;
   private ServerMetrics _serverMetrics;
+  boolean _enableThreadCpuTimeInstrument;
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:80]: This can be removed

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  public void stop() {
+    if (IS_THREAD_CPU_TIME_MEASUREMENT_ENABLED) {
+      _endTimeNs = MX_BEAN.getCurrentThreadCpuTime();
+    }
+  }
+
+  public long getThreadTimeNs() {
+    if (_startTimeNs == -1 || _endTimeNs == -1) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:58]: Do we need this if check? What is the worst that can happen if we call `get` without calling start? I vote for avoiding this. Instead if thread cpu timing is not enabled, return 0.

Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/ThreadTimer.java:58]: Will send another PR to address this.

File: pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java (hunk 20 above)
Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/query/executor/ServerQueryExecutorV1Impl.java:80]: Will send another PR to address this.

File: pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java (hunk 18 above)
Comment (mqliang) [reply pinot-core/src/test/java/org/apache/pinot/core/operator/combine/CombineSlowOperatorsTest.java:68]: Oh, this is performed by IntellJ reformat during rebase. Since it's a common practice to limit the length of line and this PR already get merged, I vote to not undo this change.
